File: backend/outputs.py
"""
Universal output recorder. Writes to the `outputs` table for every AI-generated
output in Signalera. This is the foundation for feedback collection, outcome
grading, and cross-output learning.
"""
from __future__ import annotations

import logging
from typing import Any, Optional, Literal
from uuid import UUID

from supabase import Client

logger = logging.getLogger(__name__)

# Every value this codebase is allowed to write to `outputs.output_type`.
#
# THE THIRD LIST. There are three statements of one fact:
#
#   A. this Literal                        written by a developer editing Python
#   B. OUTPUT_TYPES in src/lib/outputs.ts  written by a developer editing TS
#   C. public.output_type_enum             written by Postgres, and the only one
#                                          that decides anything at runtime
#
# A and B are claims ABOUT C. Postgres rejects a non-member with SQLSTATE 22P02
# and the supabase client surfaces that without raising, so a list that runs
# ahead of the enum fails silently, forever, at runtime only. That is exactly
# what 'company_overview' did on the TypeScript side.
#
# This list had drifted to 14 members against an enum of 16, with nothing
# comparing them. It is now pinned to tests/fixtures/output-type-enum.json,
# which is captured FROM the database by scripts/capture-output-type-enum.mjs
# and is not hand-authored. backend/tests/test_output_type_enum.py enforces it,
# and enforces separately that every output_type any backend module actually
# passes to record_output is a member here. Adding a value requires a migration
# that adds the same value to output_type_enum, declared under "pending" in that
# fixture.
OutputType = Literal[
    'memo', 'brief', 'brief_cluster', 'brief_section', 'chat_answer',
    'thesis', 'thesis_grade', 'contrarian_signal', 'deal_extraction',
    'user_addendum', 'mention_alert', 'cross_reference',
    'sec_filing', 'insider_transaction',
    'radar_clusters', 'radar_cluster_label',
    'company_overview',
]


def record_output(
    supabase: Client,
    *,
    output_type: OutputType,
    content: dict[str, Any],
    generation_context: Optional[dict[str, Any]] = None,
    user_id: Optional[str | UUID] = None,
    source_table: Optional[str] = None,
    source_id: Optional[str | UUID] = None,
) -> Optional[str]:
    """
    Record a generated output. Returns the output id (UUID string) on success, None on failure.
    Failures are logged but do not raise — recording must never block generation.
    """
    try:
        payload: dict[str, Any] = {
            'output_type': output_type,
            'content': content,
            'generation_context': generation_context or {},
        }
        if user_id is not None:
            payload['user_id'] = str(user_id)
        if source_table is not None:
            payload['source_table'] = source_table
        if source_id is not None:
            payload['source_id'] = str(source_id)

        result = supabase.table('outputs').insert(payload).execute()
        if result.data and len(result.data) > 0:
            return result.data[0]['id']
        return None
    except Exception as e:
        logger.error("[outputs] Failed to record %s: %s", output_type, e)
        return None


def record_outputs_batch(
    supabase: Client,
    rows: list[dict[str, Any]],
) -> list[Optional[str]]:
    """
    Batch-insert multiple output rows. Used for brief clusters where one brief
    generation produces many output rows.
    Each dict in `rows` should have the same shape as the kwargs of record_output.
    Returns list of output ids (or None for any that failed) in the same order.
    """
    if not rows:
        return []
    try:
        payloads = []
        for r in rows:
            p: dict[str, Any] = {
                'output_type': r['output_type'],
                'content': r['content'],
                'generation_context': r.get('generation_context') or {},
            }
            if r.get('user_id') is not None:
                p['user_id'] = str(r['user_id'])
            if r.get('source_table') is not None:
                p['source_table'] = r['source_table']
            if r.get('source_id') is not None:
                p['source_id'] = str(r['source_id'])
            payloads.append(p)

        result = supabase.table('outputs').insert(payloads).execute()
        if result.data:
            return [row['id'] for row in result.data]
        return [None] * len(rows)
    except Exception as e:
        logger.error("[outputs] Batch insert failed (%d rows): %s", len(rows), e)
        return [None] * len(rows)

File: src/lib/outputs.ts
/**
 * Universal output recorder. Writes to the `outputs` table for every
 * AI-generated output. Foundation for feedback collection and outcome grading.
 */
import type { SupabaseClient } from '@supabase/supabase-js';

/**
 * Every value this codebase is allowed to write to `outputs.output_type`.
 *
 * THIS LIST IS A CLAIM ABOUT THE DATABASE, NOT A LOCAL PREFERENCE. The column
 * is `public.output_type_enum`. Postgres rejects any value that is not a member
 * with SQLSTATE 22P02, and supabase-js reports that in the result object rather
 * than throwing, so an unbacked entry here fails silently at runtime forever.
 *
 * That is exactly what happened to 'company_overview': it was added here with
 * no matching enum member, so every Coverage Primer cache write 22P02'd and
 * every page view re-billed a Gemini call.
 *
 * THREE LISTS, ONE FACT. This is not the only statement of it:
 *
 *   A. this constant                     written by a developer editing TS
 *   B. `OutputType` in backend/outputs.py written by a developer editing Python
 *   C. `public.output_type_enum`          written by Postgres, and the only one
 *                                         that decides anything at runtime
 *
 * A and B had drifted to 15 and 14 members against an enum of 16. Both are now
 * pinned to tests/fixtures/output-type-enum.json, which is captured FROM the
 * database by scripts/capture-output-type-enum.mjs and is not hand-authored:
 * src/lib/outputs.enum.test.ts guards A, backend/tests/test_output_type_enum.py
 * guards B. They are held EQUAL to the enum rather than merely contained by it,
 * so there is one true statement of what the column accepts in each language
 * instead of three partial ones.
 *
 * Adding an entry here REQUIRES a migration that adds the same value to
 * output_type_enum, declared under "pending" in that fixture.
 */
export const OUTPUT_TYPES = [
  'memo',
  'brief',
  'brief_section',
  'brief_cluster',
  'chat_answer',
  'thesis',
  'thesis_grade',
  'contrarian_signal',
  'deal_extraction',
  'user_addendum',
  'mention_alert',
  'cross_reference',
  'sec_filing',
  'insider_transaction',
  'radar_clusters',
  'radar_cluster_label',
  'company_overview',
] as const;

export type OutputType = (typeof OUTPUT_TYPES)[number];

/**
 * Runtime membership check for a value that did not come from TypeScript.
 *
 * `OutputType` is erased at compile time, so any route that types a request
 * body as `{ output_type: OutputType }` is asserting, not checking. A caller
 * can POST any string at all and it goes straight to Postgres, which 22P02s it
 * into a log. Anything reading an output_type off the wire must run it through
 * here first.
 */
export function isOutputType(value: unknown): value is OutputType {
  return typeof value === "string" && (OUTPUT_TYPES as readonly string[]).includes(value);
}

/**
 * `outputs.source_id` is a `uuid` column, not text.
 *
 * Handing it anything else raises 22P02 ("invalid input syntax for type uuid")
 * and takes the WHOLE insert down with it, which is how the Coverage Primer
 * cache lost every row twice over: once on the enum and once, independently, on
 * a company NAME passed as source_id. The parameter is typed `string`, so
 * TypeScript is no help. Checked at the boundary instead.
 */
const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export function isUuid(value: unknown): value is string {
  return typeof value === "string" && UUID_RE.test(value);
}

interface RecordOutputParams {
  output_type: OutputType;
  content: Record<string, unknown>;
  generation_context?: Record<string, unknown>;
  user_id?: string | null;
  source_table?: string | null;
  source_id?: string | null;
}

/**
 * Record a generated output. Returns output id (UUID) on success, null on failure.
 *
 * Failures are logged but never thrown - recording must not block generation.
 * That contract is deliberate and is kept: several callers record on a path
 * where a throw would break the user-facing response.
 *
 * The cost of that contract is that `null` is the ONLY failure signal a caller
 * gets. Callers that depend on the write landing (a cache, for example) must
 * check the return value; ignoring it is how a permanently failing write stays
 * invisible. The log line now carries the SQLSTATE code so 22P02-class schema
 * mismatches are identifiable from a log body instead of just "failed".
 */
export async function recordOutput(
  supabase: SupabaseClient,
  params: RecordOutputParams
): Promise<string | null> {
  try {
    // Reject the two 22P02 shapes at the boundary rather than at the database.
    // The outcome is `null` either way, so this changes no caller's behaviour;
    // what it changes is that the log names the real cause instead of an opaque
    // SQLSTATE arriving from a round trip nobody watches.
    if (!isOutputType(params.output_type)) {
      console.error(
        "[outputs] Refusing to record an output_type the enum will reject:",
        JSON.stringify({ output_type: params.output_type, allowed: OUTPUT_TYPES })
      );
      return null;
    }
    if (params.source_id != null && !isUuid(params.source_id)) {
      console.error(
        `[outputs] Refusing to record ${params.output_type}: source_id is a uuid column and this is not a uuid:`,
        JSON.stringify({ source_id: params.source_id })
      );
      return null;
    }

    const payload: Record<string, unknown> = {
      output_type: params.output_type,
      content: params.content,
      generation_context: params.generation_context ?? {},
    };
    if (params.user_id) payload.user_id = params.user_id;
    if (params.source_table) payload.source_table = params.source_table;
    if (params.source_id) payload.source_id = params.source_id;

    const { data, error } = await supabase
      .from('outputs')
      .insert(payload)
      .select('id')
      .single();

    if (error) {
      console.error(
        `[outputs] Failed to record ${params.output_type}:`,
        JSON.stringify({ code: error.code, message: error.message, details: error.details })
      );
      return null;
    }
    return data?.id ?? null;
  } catch (e) {
    console.error(`[outputs] Exception recording ${params.output_type}:`, e);
    return null;
  }
}
